Diversify GitHub CI build platforms: add macOS 11, Windows 2022

tautschnig opened the pull request:
GitHub runners have gained support for Windows 2022 and macOS 11. Switch
some of our CI jobs to these newer platforms to make sure we detect
platform-specific issues. There remain other CI jobs on the previous
platforms (Windows 2019 and macOS 10.15, respectively).

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Diversify GitHub CI build platforms: add macOS 11, Windows 2022

GitHub runners have gained support for Windows 2022 and macOS 11. Switch
some of our CI jobs to these newer platforms to make sure we detect
platform-specific issues. There remain other CI jobs on the previous
platforms (Windows 2019 and macOS 10.15, respectively).

tautschnig commented:
Turns out some updates to our vector built-ins are required for regressions to pass on MacOS 11.

NlightNFotis reviewed (commented):
LGTM, with the caveat of the tests passing issue.

tautschnig committed:
Clang intrinsics: add declarations of AMX built-ins

Intel Advanced Matrix Extensions (AMX) yield another set of intrinsics
to be supported by compilers. Function declarations added based on
https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Basic/BuiltinsX86_64.def.

NlightNFotis commented:
@tautschnig You may have to force push again (rebase on develop, or change commit timestamps so that it resets the commit id) because two of the jobs have bugged (they haven't started running yet, and when this happens they never do).

tautschnig commented:
> @tautschnig You may have to force push again (rebase on develop, or change commit timestamps so that it resets the commit id) because two of the jobs have bugged (they haven't started running yet, and when this happens they never do).

Yes, I'll do so once https://github.com/diffblue/cbmc/pull/6642 has completed CI and been merged. Whatever the issue, I'm not too worried about those two jobs as they weren't touched in this PR.

chris-ryder commented on .github/workflows/pull-request-checks.yaml:
```diff
@@ -426,8 +426,8 @@ jobs:
 
-  check-vs-2019-make-build-and-test:
-    runs-on: windows-2019
+  check-vs-2022-make-build-and-test:
```
@peterschrammel Are there any downstream users of CBMC that we still need to support on VS 2019?

peterschrammel replied:
Very likely, checking...

peterschrammel replied:
Yes, there are. I think it's fine to upgrade the make job to VS2022 as long as we keep the VS2019 cmake job.

tautschnig commented:
@NlightNFotis or whoever has such permissions: for this PR to pass the set of "Required" CI checks will need to be updated first, perhaps with a temporary update: the two checks marked "Expected" are those that I've replaced in this PR by ones with a newer platform (and thought that corresponding updates to their names were a good idea).

NlightNFotis commented:
Hi Michael, I think @peterschrammel has administration rights on the repository, I'll chase it up with him. 

Thanks for the update.

NlightNFotis reviewed (approved):
LGTM (update to my review)
